Storage/REST-API.md

```diff
@@ -0,0 +1,185 @@
+
+```
+{
+  “error”: “Detailed human-readable error message.”,
+  ... other fields ...
+}
+```
+
+Successful responses are JSON objects without the top-level `"error"` field.
+
+`TODO: We also may have redirects, which contain the "Location" in the header, with empty body.`
+
+### Notation
+
+The default way to encode JSON objects is JavaScript-friendly. The API also supports F#-friendly format.
+
+The differences are:
+
+1. Variant types.
+
+   If a field `x` can be either an `apple` or an `orange`, the JavaScript-friendly representation is `{...,"x":{"apple":{...},...}`, and the F#-friendly representation is `{...,"x":{"Case":"apple","Fields":[...]},...}}`.
+
+2. Coding of absent optional objects.
+
+   The JavaScript-friendly implementation omits missing fields from the output (and does not expect them in the input).
+The F#-friendly implementation outputs missing fields as `null`-s (and expects `null`-s in the input).
+
+3. Coding of dictionaries with the keys being strings.
+
+   Generally, both JavaScript- and F#-friendly implementations encode dictionaries as arrays of arrays of two elements: `[[key1,value1],[key2,value2],...]`.
+   
+   However, if the type of `key` is string, the JavaScript-friendly format of this dictionary becomes an object: `{"key1":value1,"key2":value2,...}`.
+
+   `TODO(dkorolev): Check if some other F#-related issues should be added into this section.`
+
+### Views
+
+**View** is a representation of a record, which contains some non-empty subset of the record’s fields.
+The API supports different views via the `&format=...` URL parameter.
+
+By default, records returned by the API contain all fields.
+
+Two implicitly defined views (both default to "all fields") are `full` and `preview`.
+
+The `full` view is the default, so that `&format=full` is equivalent to not specifying the `&format` URL parameter.
+
+The `&format=preview` is used when the response schema for certain endpoint contains not one record, but a collection of them. In other words, the representation of each record of `"data":[...]` for the endpoints where `"data"` is an array is equivalent to the representation of those records with `&format=preview`.
+
+The views are defined by the user, and the default representation of any view is "all fields". In particular, unless the user defines the `"preview"` view, it will be identical to the `"full"` one.
+
+### Methods
+
+Current Storage API interprets HTTP verbs in the following way:
+
+* **GET**: Get the resource.
+
+  Strongly typed with respect to returning an individual record or a collection of records. For example, `/users` can be the way to GET the list of users (`"data":[...]`), and `/user/1` can be the way to GET an individual user with the ID of `1`.
+   
+* **PATCH**: Update some fields of the resource.
+
+  `PATCH` should be called on the URL that has been (or could have been) returned as the top-level `"url":"..."` field of the payload of `GET`.
```

Comment:
Good point. Allowing to `PUT/PATCH/DELETE` only the resource containing the most recent `&timestamp` in the URL is the most elegant way we found to ensure no "merge conflicts" when multiple clients are accessing the same resource.

Suggestions for a better solution would be greatly appreciated. Thank you!